lib/spack/spack/directives.py

```diff
@@ -330,30 +336,35 @@ def version(ver, checksum=None, **kwargs):
     Keyword Arguments:
         deprecated (bool): whether or not this version is deprecated
     """
+    return lambda pkg: _execute_version(pkg, ver, checksum, **kwargs)
 
-    def _execute_version(pkg):
-        if checksum is not None:
-            if hasattr(pkg, "has_code") and not pkg.has_code:
-                raise VersionChecksumError(
-                    "{0}: Checksums not allowed in no-code packages"
-                    "(see '{1}' version).".format(pkg.name, ver)
-                )
 
-            kwargs["checksum"] = checksum
-
-        # Store kwargs for the package to later with a fetch_strategy.
-        version = Version(ver)
-        if isinstance(version, GitVersion):
-            if not hasattr(pkg, "git") and "git" not in kwargs:
-                msg = "Spack version directives cannot include git hashes fetched from"
-                msg += " URLs. Error in package '%s'\n" % pkg.name
-                msg += "    version('%s', " % version.string
-                msg += ", ".join("%s='%s'" % (argname, value) for argname, value in kwargs.items())
-                msg += ")"
-                raise VersionLookupError(msg)
-        pkg.versions[version] = kwargs
-
-    return _execute_version
+def _execute_version(pkg, ver: Union[str, int], checksum=None, **kwargs):
```

Comment:
Any particular reason not to keep this as an internal closure of `version()`?  Oh I see this is to make it more testable?  That seems like a good reason.

Reply:
Yeah, I just wanted to pass a `pkg` as namedtuple so that it all fits in a single test w/o a new mock package